### ISSUE TITLE
Sort lines sorts quoted values inconsistently

### ISSUE THREAD

**Issue by AnthonyMastrean** [microsoft/vscode#60978]
Issue Type: <b>Bug</b>

Sort lines sorts quoted values inconsistently with the same values without double-quotes. I searched for [similar bugs](https://github.com/Microsoft/vscode/issues?q=is%3Aissue+is%3Aopen+%22sort+lines%22+label%3Abug) but didn't find much. Although, there's some complaint about [sorting with mixed case](https://github.com/Microsoft/vscode/issues/18315#issuecomment-352255186), but I specifically avoid that in my examples.

![vscode-sort-lines](https://user-images.githubusercontent.com/217842/46959997-70f54380-d06b-11e8-9c3a-cf2002ca5522.gif)

Sort the following lines (asc)

```
a
abc-def
ab
abc
```

You should get

```
a
ab
abc
abc-def
```

Now sort the same lines, but wrapped in double-quotes (like items in a JSON literal often are)

```
"a"
"abc-def"
"ab"
"abc"
```

I expect to get

```
"a"
"ab"
"abc"
"abc-def"
```

But I actually get this, which _must_ be incorrect, right?

```
"a"
"ab"
"abc-def"
"abc"
```

VS Code version: Code 1.28.0 (431ef9da3cf88a7e164f9d33bf62695e07c6c2a9, 2018-10-05T14:58:53.203Z)
OS version: Windows_NT x64 10.0.17134

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-6600U CPU @ 2.60GHz (4 x 2808)|
|GPU Status|2d_canvas: enabled<br>checker_imaging: disabled_off<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>native_gpu_memory_buffers: disabled_software<br>rasterization: enabled<br>video_decode: enabled<br>video_encode: enabled<br>webgl: enabled<br>webgl2: enabled|
|Memory (System)|15.89GB (2.21GB free)|
|Process Argv|.|
|Screen Reader|no|
|VM|0%|

</details><details><summary>Extensions (7)</summary>

Extension|Author (truncated)|Version
---|---|---
tslint|eg2|1.0.40
vscode-gitk|how|1.3.2
csharp|ms-|1.16.2
Go|ms-|0.6.91
PowerShell|ms-|1.9.0
vsliveshare|ms-|0.3.790
vscode-docker|Pet|0.3.1

(1 theme extensions excluded)

</details>

**Comment by irrationalRock** [comment]
Hi, I have looked into this issue and I believe I found why the sorting is not consistent. If you look at https://github.com/Microsoft/vscode/blob/43823dae97f07760958ffbb5e16c7f219307159c/src/vs/editor/contrib/linesOperations/sortLinesCommand.ts#L75  it uses `localeCompare`. `localeCompare` does not regonize `"a"` and `a` as the same.  This can be demonstrated with this line of code:
```javascript
var a = '\"A\"'; 
var b = 'A'; 
console.log(a.localeCompare(b));
``` 
I think one way of solving the problem would be to strip all special characters from the string array and then sorting.

**Comment by rebornix** [comment]
IMO the root cause is we try to compare `-` and `"` and `-` wins in this case.

**Comment by actualben** [comment]
TLDR; impact of unicode collation explored

This particular example does look very strange! Here are the [code point](https://en.wikipedia.org/wiki/Code_point)s for the characters in these strings:

```javascript
let unquoted_strings = [
    [97],  // a
    [97, 98],  // ab
    [97, 98, 99],  // abc
    [97, 98, 99, 45, 100, 101, 102],  // abc-def
]

let quoted_strings = [
    [34, 97, 34],  // "a"
    [34, 97, 98, 34],  // "ab"
    [34, 97, 98, 99, 45, 100, 101, 102, 34],  // "abc-def"
    [34, 97, 98, 99, 34],  // "abc"
]
```

...It makes no apparent sense that 34 sorts after 45 in those last two strings... and there's this:

```javascript
'a' < 'b' // true
// in compare functions: -1 means a<b; 0 means a==b; 1 means a>b
'a'.localeCompare('b') // -1 
'"' < '-' // true
'"'.localeCompare('-') // 1 <-- ??? not -1 ??? (╯°□°) ╯ ︵┻━┻
```

Why? Because Javascript's locale-aware (unicode) string comparison function (used by vscode sort functions) is not primarily based on comparing code point values, instead it is done ["by assigning multi-level weights to characters or sequences of characters and then comparing those weights on each level"](https://unicode.org/faq/collation.html#0) aka "collation."

> the standard Unicode Collation Algorithm (UCA) comes with a default weight table for all assigned characters

I believe that refers to the [latest UCA allkeys file](https://www.unicode.org/Public/UCA/latest/allkeys.txt) -- the [format of the allkeys file is described here](https://www.unicode.org/reports/tr10/#File_Format).

Here are some lines from the current allkeys file (in no particular order) :
```
0022  ; [*031D.0020.0002] # QUOTATION MARK
002D  ; [*020D.0020.0002] # HYPHEN-MINUS
0061  ; [.1FA2.0020.0002] # LATIN SMALL LETTER A
0062  ; [.1FBC.0020.0002] # LATIN SMALL LETTER B
0063  ; [.1FD6.0020.0002] # LATIN SMALL LETTER C
```

Expressing the above hexadecimal numbers in base 10:
```
34  ; [*797.32.2] # QUOTATION MARK
45  ; [*525.32.2] # HYPHEN-MINUS
97  ; [.8098.32.2] # LATIN SMALL LETTER A
98  ; [.8124.32.2] # LATIN SMALL LETTER B
99  ; [.8150.32.2] # LATIN SMALL LETTER C
```

So for the codepoints 34 (quotation mark) and 45 (hyphen/minus) their respective weights boil down to 797 and 525 respectively. They are kinda "flipped" at least with respect to ASCII/their codepoint values. `(╯°□°) ╯ ︵ᴉᴉɔsɐ`

What could be done to make vscode sort in the way you might expect/want? `localeCompare` supports some options like `ignorePunctuation` (and also `sensitivity`)

```javascript
let quoted_strings = [ '"a"', '"ab"', '"abc-def"',  '"abc"' ];
quoted_strings.sort(
  (a,b)=> a.localeCompare(b, undefined, {ignorePunctuation: true})
);
// quoted_strings == [ '"a"', '"ab"', '"abc"', '"abc-def"' ]
```

Setting `ignorePunctuation: true` would help this case, but would be very unhelpful for others. I would love to just have more sort commands in the menu with different values for `ignorePunctuation` and `sensitivity` which give access to the sorting we need for a particular purpose. It could also be useful to have a sort command that does actually sort by codepoint values.
